pkg/pubsub/pubsub.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2023 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package pubsub provides an interface and implementations for publishing
+// notifications for Rekor updates to a Pub/Sub system.
+package pubsub
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+)
+
+// Publisher provides methods for publishing notifications of new entries in the
+// transparency log.
+type Publisher interface {
+	// Publish publishes a transparency log entry to the configured Pub/Sub topic.
+	Publish(ctx context.Context, entry models.LogEntryAnon) error
+	// Close safely closes any active connections.
+	Close() error
+}
+
+// ProviderNotFoundError indicates that no matching PubSub provider was found
+type ProviderNotFoundError struct {
+	ref string
+}
+
+func (e *ProviderNotFoundError) Error() string {
+	return fmt.Sprintf("no pubsub provider found for key reference: %s", e.ref)
+}
+
+// ProviderInit is a function that initializes provider-specific Publisher.
+type ProviderInit func(ctx context.Context, topicResourceID string) (Publisher, error)
+
+// AddProvider adds the provider implementation into the local cache
+func AddProvider(uri string, init ProviderInit) {
+	providersMap[uri] = init
+}
+
+var providersMap = map[string]ProviderInit{}
+
+// Get returns a Publisher for the given resource string and hash function.
+// If no matching provider is found, Get returns a ProviderNotFoundError. It
+// also returns an error if initializing the Publisher fails. If no resource
+// is supplied, it returns a nil Publisher and no error.
+func Get(ctx context.Context, topicResourceID string) (Publisher, error) {
+	if topicResourceID == "" {
+		return nil, nil
```

Comment:
should we not warn or err if a topicResourceID is blank?

Reply:
It should not error. Publishing notifications is an optional feature that operators of private instances (or some e2e test scenarios) may not want to enable.

I'm happy to add a warning log, will that work?

Reply:
I think you could wrap the code in [pkg/api/api.go](https://github.com/sigstore/rekor/pull/1561/files/472860e355c486235f402ce1e3f050d75e5c325f#diff-67a6e36014be865053c43d38b2091bd2d3f17a91b117d97c0f5244cd7d71127d) with a `viper.IsSet("rekor_server.new_entry_publisher")` check. 

You could then return an error if `topicResourceID` is an empty string as @lukehinds is suggesting. I think "returning an error if `topicResourceID` is empty" is more consistent with the coding style we've used elsewhere in sigstore.

Reply:
Ack, done.